src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go

```diff
@@ -287,39 +292,58 @@ func (d *SyncServerDatabase) CompleteSync(
 	}
 
 	// Build up a /sync response. Add joined rooms.
-	res := types.NewResponse(pos)
+	res := types.NewResponse(posTo)
 	for _, roomID := range roomIDs {
-		var stateEvents []gomatrixserverlib.Event
-		stateEvents, err = d.roomstate.selectCurrentState(ctx, txn, roomID)
-		if err != nil {
-			return nil, err
-		}
-		// TODO: When filters are added, we may need to call this multiple times to get enough events.
-		//       See: https://github.com/matrix-org/synapse/blob/v0.19.3/synapse/handlers/sync.py#L316
-		var recentStreamEvents []streamEvent
-		recentStreamEvents, err = d.events.selectRecentEvents(
-			ctx, txn, roomID, types.StreamPosition(0), pos, numRecentEventsPerRoom,
-		)
-		if err != nil {
-			return nil, err
+
+		jr := types.NewJoinResponse()
+
+		//Join response should contain events only if room isn't filtered
+		if !isRoomFiltered(roomID, filter, nil) {
```

Comment:
AFAIK a wished structure for this kind of checks would be in this case
```
if isRoomFiltered(roomID, filter, nil) {
    continue
}
```
instead of having a closing backet dozens of lines later

Reply:
Make sense, however each joined room must appear in the `rooms.join` list, even if the room has been filtered.

In fact, the room filtering only hides the events, but keep everything else as is.

This is an example of a join response for a filtered room on matrix.org:
```json
"!TUYoaUpGiPzdpBzfVM:matrix.org": {
	"unread_notifications": {
		"highlight_count": 0,
		"notification_count": 0
	},
	"timeline": {
		"limited": false,
		"prev_batch": "s419663949_378214671_17641_89713136_35031270_231422_3761477_4179932_4347",
		"events": []
	},
	"state": {
		"events": []
	},
	"ephemeral": {
		"events": []
	},
	"account_data": {
		"events": [
			{
				"content": {
					"primary_color": "#e55e5e",
					"secondary_color": "#f5eaea"
				},
				"type": "org.matrix.room.color_scheme"
			},
			{
				"content": {
					"event_id": "$15162674503862115fsbXn:matrix.org"
				},
				"type": "m.fully_read"
			}
		]
	}
},
```